File: go/elliptic/elliptic.go
package elliptic

import "fmt"

func Abs(n int,base int) int {
	if n < 0 {
		return n%base+base
	}
	return n%base
}

func Expo(x int, y int, base int) int {
	x_, y_ := Abs(x,base), Abs(y,base)
	if y_ == 0 {
		return 1
	}
	xp := Expo(x_, y_/2, base)
	if y_%2 == 0 {
		return xp * xp % base
	}
	return Abs(x_ * xp * xp, base)
}

func Inverse(n int, base int) (int, error) {
	n_ := Abs(n,base)
	if n_ == 0 {
		return 0, fmt.Errorf("elliptic: divided by zero modulo %d", base)
	}
	return Expo(n_, base-2, base), nil
}

func Divide(n int, q int, base int) (int, error) {
	i,e := Inverse(q, base)
	if e != nil {
		return 0,e
	}
	return Abs(n * i, base),nil
}

var base int

func Base() int {
	return base
}

func SetBase(b int) {
	base = b
}

type Point struct {
	X, Y, Z int
}

func NewPoint(x int, y int, z int) *Point {
	return &Point{X: Abs(x,base), Y: Abs(y,base), Z: Abs(z,base)}
}

func (p *Point) Equals(q *Point) bool {
	return p.X == q.X && p.Y == q.Y && p.Z == q.Z
}

type ProjPoint struct {
	Point
}

func NewProjPoint(c ...int) (*ProjPoint, error) {
	if len(c) < 2 {
		return nil, fmt.Errorf("elliptic: a projpoint needs X and Y")
	}
	var z int
	if len(c) < 3 {
		z = 1
	} else {
		z = c[2]
	}
	p := NewPoint(c[0], c[1], z)
	if (p.X + p.Y + p.Z) == 0 {
		return nil, fmt.Errorf("elliptic: point (0,0,0) is not projective")
	}
	return &ProjPoint{*p},nil
}

func (p *ProjPoint) IsInfty() bool {
	return p.Z == 0
}

func (p *ProjPoint) Equals(q *ProjPoint) (bool,error) {
	if p.IsInfty() {
		if !q.IsInfty() {
			return false,nil
		}
		if p.X == 0 {
			return q.X == 0,nil
		}
		d,e := Divide(q.X, p.X, base)
		if e != nil {
			return false,e
		}
		return d*p.Y%base == q.Y,nil
	}
	if q.IsInfty() {
		return false,nil
	}
	return p.X == q.X && p.Y == q.Y,nil
}

func (p *ProjPoint) HX() (int, error) {
	if p.IsInfty() {
		return 0, fmt.Errorf("elliptic: infinity cannot be homogeneous")
	}
	return Divide(p.X, p.Z, base)
}

func (p *ProjPoint) HY() (int, error) {
	if p.IsInfty() {
		return 0, fmt.Errorf("elliptic: infinity cannot be homogeneous")
	}
	return Divide(p.Y, p.Z, base)
}

func (p *ProjPoint) BelongsTo(a int, b int) bool {
	if p.IsInfty() {
		return p.X == 0 && p.Y == 1 && p.Z == 0
	}
	return (p.X*p.X*p.X+a*p.X*p.Z*p.Z+b*p.Z*p.Z*p.Z-p.Y*p.Y*p.Z)%base == 0
}

var a, b int

func A() int {
	return a
}

func SetA(a_ int) {
	a = a_
}

func B() int {
	return b
}

func SetB(b_ int) {
	b = b_
}

type ElliPoint struct {
	ProjPoint
}

func NewElliPoint(c ...int) (*ElliPoint, error) {
	p, e := NewProjPoint(c...)
	if e != nil {
		return nil, fmt.Errorf("elliptic: invalid projection\n%s", e)
	}
	if !p.BelongsTo(a, b) {
		return nil, fmt.Errorf("elliptic: point out of curve (%d,%d)", a, b)
	}
	return &ElliPoint{*p},nil
}

var ElliInfty *ElliPoint

func init() {
	ElliInfty = &ElliPoint{ProjPoint{Point{X: 0, Y: 1, Z: 0}}}
}

func (p *ElliPoint) Opp() *ElliPoint {
	if p.IsInfty() {
		return ElliInfty
	}
	x,_ := p.HX()
	y,_ := p.HY()
	q,_ := NewElliPoint(x,-y%base)
	return q
}

func (p *ElliPoint) Star(q *ElliPoint) *ElliPoint {
	if p.IsInfty() {
		return q.Opp()
	}
	if q.IsInfty() {
		return p.Opp()
	}
	x,_ := p.HX()
	y,_ := p.HY()
	var k, l int
	equal,_ :=p.Equals(&q.ProjPoint)
	if equal {
		l,_ = Divide(3*x*x+a, 2*y, base)
		k = l*l - 2*x
		r,_ := NewElliPoint(k, l*(k-x)+y)
		return r
	}
	x_,_ := q.HX()
	y_,_ := q.HY()
	if x == x_ {
		return ElliInfty
	}
	l,_ = Divide(y_-y, x_-x, base)
	k = l*l - x - x_
	r,_ := NewElliPoint(k, l*(k-x)+y)
	return r
}

func (p *ElliPoint) Plus(q *ElliPoint) *ElliPoint {
	return ElliInfty.Star(p.Star(q))
}

func (p *ElliPoint) Minus(q *ElliPoint) *ElliPoint {
	return p.Plus(q.Opp())
}

func (p *ElliPoint) Times(n int) *ElliPoint {
	if n == 0 {
		return ElliInfty
	}
	if n < 0 {
		return p.Opp().Times(-n)
	}
	return p.Plus(p.Times(n - 1))
}

File: python/elliptic/elliptic.py
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see [http://www.gnu.org/licenses/].
#
# Copyright (C) 2013 Nicolas Canceill
#

def expo(x,y,base):
	x_ = x%base
	y_ = y%base
	if y_ == 0: return 1
	if y_%2 == 0: return expo(x_,y_//2,base)**2
	return x_*expo(x_,y_-1,base)

def inverse(n,base):
	n_ = n%base
	if n_ == 0:
		raise ZeroDivisionError("divided by 0 modulo "+str(base))
	return expo(n_,base-2,base)

def divide(n,q,base):
	return n * inverse(q,base) % base

class Point:
	base = 0

	def __init__(self,x,y,z):
		self.X = x % self.base
		self.Y = y % self.base
		self.Z = z % self.base

	def eq(self,p):
		return self.X == p.X and self.Y == p.Y and self.Z == p.Z

	def __str__(self):
		return "Point ({},{},{})".format(self.X,self.Y,self.Z)

class ProjPoint(Point):
	def __init__(self,x,y,z=1):
		super().__init__(x,y,z)
		if self.X+self.Y+self.Z == 0:
			raise TypeError("(0,0,0) is not projective")

	def is_infty(self):
		return self.Z % self.base == 0

	def eq(self,p):
		if self.is_infty():
			if not p.is_infty(): return false
			if self.X == 0: return p.X == 0
			return divide(p.X,self.X,self.base) * self.Y % base == p.Y
		else:
			if p.is_infty(): return false
			return self.X == p.X and self.Y == p.Y

	def x(self):
		if self.is_infty():
			raise TypeError("Infinity cannot be homogeneous")
		return divide(self.X,self.Z,self.base)

	def y(self):
		if self.is_infty():
			raise TypeError("Infinity cannot be homogeneous")
		return divide(self.Y,self.Z,self.base)

	def __str__(self):
		if self.is_infty(): return "ProjPoint Infinity/{}".format(self.base)
		return "ProjPoint ({},{})/{}".format(self.x(),self.y(),self.base)

	def belongs_to(self,a,b):
		if self.is_infty(): return self.X == 0 and self.Y == 1
		return (self.X**3+a*self.X*self.Z**2+b*self.Z**3-self.Y**2*self.Z)%self.base == 0

class ElliPoint(ProjPoint):

	a = 0
	b = 0

	@staticmethod
	def infty():
		return ElliPoint(0,1,0)

	@classmethod
	def curve_set(self,a,b):
		self.a = a
		self.b = b

	def __init__(self,x,y,z=1):
		super().__init__(x,y,z)
		if not self.belongs_to(self.a,self.b):
			raise AttributeError("Point out of curve: {}".format(self))

	def __neg__(self):
		if self.is_infty(): return self.infty()
		return ElliPoint(self.x(),-self.y())

	def __pow__(self,p):
		if self.is_infty(): return -p
		if p.is_infty(): return -self
		if self == p:
			l = divide(3*self.x()**2+self.a,2*self.y(),self.base)
			k = l**2-2*self.x()
			return ElliPoint(k,l*(k-self.x())+self.y())
		if self.x() == p.x(): return self.infty()
		l = divide(p.y()-self.y(),p.x()-self.x(),self.base)
		k = l**2-self.x()-p.x()
		return ElliPoint(k,l*(k-self.x())+self.y())

	def __rpow__(self,p):
		return p.__pow__(self)

	def __add__(self,p):
		return self.infty()**(self**p)

	def __radd__(self,p):
		return self.__add__(p)

	def __sub__(self,p):
		return self+(-p)

	def __rsub__(self,p):
		return p.__sub__(self)

	def __mul__(self,n):
		if n == 0: return self.infty()
		if n == 1: return self
		if n < 0: return (-self)*(-n)
		return self+(self*(n-1))

	def __rmul__(self,n):
		return self.__mul__(n)

	def __str__(self):
		if self.is_infty(): return "ElliPoint Infinity/{}".format(self.base)
		return "ElliPoint ({},{})/{}".format(self.x(),self.y(),self.base)

